waspc/src/Wasp/Analyzer/TypeChecker/AST.hs

```diff
@@ -1,33 +1,50 @@
+{-# LANGUAGE DeriveFunctor #-}
+
 module Wasp.Analyzer.TypeChecker.AST
   ( TypedAST (..),
     TypedStmt (..),
     TypedExpr (..),
     Identifier,
     ExtImportName (..),
+    WithCtx (..),
+    Ctx (..),
+    withCtx,
     exprType,
   )
 where
 
-import Wasp.Analyzer.Parser (ExtImportName (..), Identifier)
+import Wasp.Analyzer.Parser (Ctx (..), ExtImportName (..), Identifier)
 import Wasp.Analyzer.Type
 
-newtype TypedAST = TypedAST {typedStmts :: [TypedStmt]} deriving (Eq, Show)
+newtype TypedAST = TypedAST {typedStmts :: [WithCtx TypedStmt]} deriving (Eq, Show)
 
-data TypedStmt = Decl Identifier TypedExpr Type deriving (Eq, Show)
+data TypedStmt
+  = -- | Decl <declName> <declBody> <declType>
+    Decl Identifier (WithCtx TypedExpr) Type
+  deriving (Eq, Show)
 
+{- ORMOLU_DISABLE -}
 data TypedExpr
-  = Dict [(Identifier, TypedExpr)] Type
-  | List [TypedExpr] Type
-  | StringLiteral String
+  = Dict           [(Identifier, WithCtx TypedExpr)] Type
+  | List           [WithCtx TypedExpr] Type
+  | StringLiteral  String
   | IntegerLiteral Integer
-  | DoubleLiteral Double
-  | BoolLiteral Bool
-  | ExtImport ExtImportName String
-  | Var Identifier Type
+  | DoubleLiteral  Double
+  | BoolLiteral    Bool
+  | ExtImport      ExtImportName String
+  | Var            Identifier Type
   | -- TODO: When adding quoters to TypeDefinitions, these JSON/PSL variants will have to be changed
-    JSON String
-  | PSL String
+    JSON           String
+  | PSL            String
   deriving (Eq, Show)
+{- ORMOLU_ENABLE -}
+
+-- TODO: Instead of defining a different WithCtx here, reuse one from Parser.AST?
+data WithCtx a = WithCtx Ctx a
+  deriving (Eq, Show, Functor)
+
+withCtx :: (Ctx -> a -> b) -> (WithCtx a -> b)
+withCtx f (WithCtx ctx expr) = f ctx expr
```

Comment:
I should reuse both `WithCtx` and `withCtx` from Parser instead of implementing our own here.

Reply:
Done